_posts/2019-10-21-Improve-performance-with-worker.md

```diff
@@ -0,0 +1,367 @@
+# 使用 Web Worker 来改善性能
+
+在 Web Worker 之前，解析 CSS，生成布局，绘制界面以及运行 javascript 脚本都运行在浏览器的一个线程里。如果一个 Web App 运行的 js 脚本一次运行时间过长，就会出现界面卡顿。这样的用户体验是没法用及格来评价的。
+
+在多核普及的当下，浏览器也大多支持了 Web Worker。这让前端开发有了更多的选择，使用 web worker 来实现真正的多线程。所有阻碍、延迟用户反馈的操作都可以移入一个后台运行的线程中。
+
+## 如何发现性能瓶颈
+
+笔者主要使用 React 开发，所以首先聊一下 React 中如何发现性能出现问题的地方。在 React16.9 中新增了 Profiler API，使用起来也非常简单，具体可以查看文档((https://reactjs.org/docs/profiler.html))。
+
+更通用一点的可以使用 chrome 的*lighthouse*。可以安装插件 lighthouse 插件，或者也可以直接打开开发面板的*audits*点击*run audits*，就能看到报表了。`console.time`和`console.timeEnd`组合。
+
+但是，以上只适用于开发模式下使用。在生产上使用多多少少会给产品本身带来额外的资源消耗。一般来说，app 都会有埋点，在埋点点时候如何顺道达成性能消耗点记录就需要具体问题具体分析了。
+
+## Web Worker
+
+使用 Web Worke 让阻塞代码在后台运行，自然不会阻塞 UI 线程（main thread）。
+
+### 创建一个 Worker
+
+创建一个 Worker 非常的简单，只需要把一段命名脚本传给`Worker`构造函数就可以。比如 MDN 的一段：
+
+这是 Worker 脚本：
+
+```js
+// worker.js
+self.onmessage = (event) => {
+  console.log('Message received', event.data);
+  self.postMessage('Worker done');
+};
+```
+
+创建 Worker：
+
+```js
+// Main thread
+var myWorker = new Worker('worker.js');
+
+myWorker.postMessage([first.value, second.value]);
+
+myWorker.onmessage = function(e) {
+  result.textContent = e.data;
+  console.log('Message received from worker');
+};
+```
+
+两个线程之间（上例是 UI thread 和一个 worker）可以通过`postMessage`和`onmessage`或者（`addEventListener('message', () => {}`）的方式来传递消息。
+
+线程之间的通信是基于**事件**的。那么错误的处理也是同样道理，例如：
+
+```js
+// UI thread
+var myWorker = new Worker('worker.js');
+
+myWorker.onerror = function() {
+  console.log('There is an error with your worker!');
+};
+```
+
+```js
+// Inside worker
+self.onerror = (err) => {
+  console.error('Error in worker', err);
+};
+```
+
+#### 引入外部脚本
+
+```js
+importScripts(); /* imports nothing */
+importScripts('foo.js'); /* imports just "foo.js" */
+importScripts('foo.js', 'bar.js'); /* imports two scripts */
+importScripts('//example.com/hello.js'); /* You can import scripts from other origins */
+```
+
+**注意**：下载顺序可以是任意顺序，但是执行的顺序是按照脚本在`importScripts`方法里出现的顺序。
+
+### 关闭一个 Worker
+
+Worker 也占用和消耗资源，所以在不用的时候就要关闭它。
+
+关闭一个 Worker 有两种方法：一种是直接在 UI thread 里面使用`terminate`方法，一种是在 Worker 的内部调用`close`方法。
+
+```javascript
+// In main thread
+const worker = new Worker('myworker.js');
+
+// If it's the time to terminate a worker
+worker.terminate();
+```
+
+在调用了`terminate`方法之后，Worker 会被立刻终止，即使是还在运行中的也是一样。但是一般情况下还是希望在 Worker 执行完成之后才去关闭。这个时候就要用到 Worker 的`close`方法。
+
+```javascript
+// In a worker
+self.onmessage = (event) => {
+  self.close();
+};
+```
+
+不过，**close**方法就要被废弃了，现在是在*deprecated*的状态。具体看 MDN 的[这里](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/close)
+
+要被废弃是因为，在一个 worker 出了作用域之后就会被回收。所以有没有`close`这个方法并没有太大的必要。
```

Comment:
That's actually not quite true.
WorkerGlobalScope is the parent to the following classes: DedicatedWorkerGlobalScope, SharedWorkerGlobalScope, and ServiceWorkerGlobalScope. Service workers do not support close(). Therefore, the implementation of close() has been moved from WorkerGlobalScope to DedicatedWorkerGlobalScope and SharedWorkerGlobalScope.

When you call new Worker, the scope class used is DedicatedWorkerGlobalScope. So you still have the close() to your disposal.